src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixCuMatMult.java

```diff
@@ -0,0 +1,361 @@
+	 * The caller is expected to invoke params.rowToColumnMajor().
+	 * 
+	 * @param handle cublas handle
+	 * @param instName name of the invoking instruction to record{@link Statistics}.
+	 * @param C output matrix pointer
+	 * @param A left matrix pointer
+	 * @param B right matrix pointer
+	 * @param param BLAS parameters
+	 * @throws DMLRuntimeException if error
+	 */
+	private static void denseDenseMatMult(cublasHandle handle, String instName, Pointer C, Pointer A, Pointer B, 
+			CuMatMultParameters param) throws DMLRuntimeException {
+		long t0 = GPUStatistics.DISPLAY_STATISTICS ? System.nanoTime() : 0;
+		String kernel = null;
+		param.rowToColumnMajor(); param.validate();
+		int transa = cublasOp(param.isLeftTransposed);
+		int transb = cublasOp(param.isRightTransposed);
+		B = swap(A, A=B);
+		if (param.m == 1 && param.n == 1){
+			// Vector product
+			LOG.debug(" GPU Dense-dense Vector Product");
+			double[] result = {0};
+			JCublas2.cublasDdot(handle, param.k, A, 1, B, 1, Pointer.to(result));
+			// By default in CuBlas V2, cublas pointer mode is set to CUBLAS_POINTER_MODE_HOST.
+			// This means that scalar values passed are on host (as opposed to on device).
+			// The result is copied from the host back to the device so that the rest of
+			// infrastructure can treat it uniformly.
+			cudaMemcpy(C, Pointer.to(result), 1 * Sizeof.DOUBLE, cudaMemcpyHostToDevice);
+			kernel = GPUInstruction.MISC_TIMER_DENSE_DOT_LIB;
+		} else if (param.m == 1) {
+			// Vector-matrix multiply
+			LOG.debug(" GPU Dense Vector-Matrix Multiply");
+			transb = reverseCublasOp(transb);
+			int rightNumRows = (transb == CUSPARSE_OPERATION_TRANSPOSE) ?  param.k : param.n;
+			int rightNumCols = (transb == CUSPARSE_OPERATION_TRANSPOSE) ? param.n : param.k;
+			JCublas2.cublasDgemv(handle, transb, rightNumRows, rightNumCols, one(), B, param.ldb, A, 1, zero(), C, 1);
+			kernel = GPUInstruction.MISC_TIMER_DENSE_VECTOR_DENSE_MATRIX_LIB;
+		} else if (param.n == 1){
+			// Matrix-vector multiply
+			LOG.debug(" GPU Dense Matrix-Vector Multiply");
+			int leftNumRows = (transa == CUSPARSE_OPERATION_NON_TRANSPOSE) ?  param.m : param.k;
+			int leftNumCols = (transa == CUSPARSE_OPERATION_NON_TRANSPOSE) ?  param.k : param.m;
+			JCublas2.cublasDgemv(handle, transa, leftNumRows, leftNumCols, one(), A, param.lda, B, 1, zero(), C, 1);
+			kernel = GPUInstruction.MISC_TIMER_DENSE_MATRIX_DENSE_VECTOR_LIB;
+		} else {
+			LOG.debug(" GPU Dense-Dense Matrix Multiply ");
+			JCublas2.cublasDgemm(handle, transa, transb, param.m, param.n, param.k, one(), A, param.lda, B, param.ldb, zero(), C, param.ldc);
+			kernel = GPUInstruction.MISC_TIMER_DENSE_MATRIX_DENSE_MATRIX_LIB;
+		}
+		if (GPUStatistics.DISPLAY_STATISTICS) GPUStatistics.maintainCPMiscTimes(instName, kernel, System.nanoTime() - t0);
+	}
+	
+	// Convenient methods to swap two values
+	// Usage: y = swap(x, x=y);
+	private static long swap(long x, long y) {
```

Comment:
Yes, it is possibly the only way to do `swap` in java. Though the `swap` function itself is difficult to read, it makes the other methods more readable instead of having 3 line swap dispersed all over the code.

Reply:
I am somewhat against this. 
In an endeavor towards brevity, we are losing clarity. In my google searches for this, I've seen this as a clever stackoverflow answer for a way to swap two primitves given that there is no concept of call by reference for primitives in java. It does not seem like an idiomatic way of doing swaps. 
A 3 line swap, or even having 3 statements on the same line would be cleaner IMHO.

Reply:
Let's agree to disagree here.